### PR TITLE
First Implementation 

### PR DESCRIPTION
**Pull request by mario-pinderi**
Jira Ticket: [PLAT-3149]

### Description
IP anonymizer implementation for IP V4 and V6.

### Types of changes

- \[ ] Bug fix
- \[x] New feature
- \[ ] Breaking change
- \[ ] Package/library update
- \[ ] Content update
- \[ ] Cosmetic
- \[ ] Documentation update

## Checklist

- \[x] My code is consistent with existing approaches of this project.
- \[ ] My change requires a change to the documentation.
- \[ ] I have updated the documentation accordingly.
- \[x] I have added tests to cover my changes.
- \[ ] The changed applications deploy successfully

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# GO IP anonymizer
+
+This library is built to give users more privacy by masking their IP address.
+
+IPv4 and IPv6 addresses are supported. 
+
+Default subnet for IPv4 is `/24` and for IPv6 is `/64`. Those can be changed when library is initialized.
+
+##Example
+```
```
Comment (fho) [inline README.md]:
```suggestion
```go
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"log"
+
+	goanonymizer "github.com/simplesurance/go-ip-anonymizer"
+)
+
+func main() {
+	ip4 := "192.168.1.12"
+	ip6 := "bbd1:e95a:adbb:b29a:e38b:577f:6f9a:1fa7"
```
Comment (fho) [inline README.md]:
```suggestion
	ip6 := "bbd1:e95a:adbb:b29a:e38b:577f:6f9a:1fa7"
	
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	newIP4, err := goanonymizer.NewAnonymize("","").AnonymizeIp(ip4)
+	if err != nil{
+		log.Println(err)
+	}
+	newIP6, err := goanonymizer.NewAnonymize("","").AnonymizeIp(ip6)
```
Comment (fho) [inline README.md]:
```suggestion

	newIP6, err := goanonymizer.NewAnonymize("","").AnonymizeIp(ip6)
```

File: anonymizer.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	v6Mask net.IPMask
+}
+
+//NewAnonymize returns instance of Anonymize.
+//If v4Mask and v6Mask are not provided, default values are used.
+func NewAnonymize(v4Mask, v6Mask string) *Anonymize {
+	ipV4 := net.ParseIP(v4Mask)
+	if ipV4 == nil {
+		ipV4 = net.ParseIP("255.255.255.0")
+	}
```
Comment (fho) [inline anonymizer.go]:
```suggestion
	}
	
```

File: anonymizer.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+	return &Anonymize{v4Mask: net.IPMask(ipV4.To4()), v6Mask: net.IPMask(ipV6)}
+}
+
+//AnonymizeIp gets IP and transform it based on v4Mask and v6Mask.
+func (a Anonymize) AnonymizeIp(ipStr string) (string, error) {
+	ip := net.ParseIP(ipStr)
+	if ip == nil {
+		return "", errors.New("IP is not valid")
+	}
```
Comment (fho) [inline anonymizer.go]:
```suggestion
	}
	
```

File: anonymizer.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+//Anonymize struct that hold mask for V4 and V6.
+type Anonymize struct {
+	v4Mask net.IPMask
+	v6Mask net.IPMask
+}
+
+//NewAnonymize returns instance of Anonymize.
+//If v4Mask and v6Mask are not provided, default values are used.
+func NewAnonymize(v4Mask, v6Mask string) *Anonymize {
```
Comment (fho) [inline anonymizer.go]: Please describe in the comment:
```suggestion
// New returns a new IP-Anonymizer.
// The v4Mask and v6Mask parameter define the host-part that are anonymized.
// For example if v4mask is 255.255.255.0, the IP address 192.168.0.5 is anonymized to 192.168.0.0.
// If v4Mask or v6Mask is empty the default values (DefaultIPv4Mask, DefaultIPv6Mask) are used.
func New(v4Mask, v6Mask string) *IPAnonymizer {
```

File: anonymizer.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package goanonymizer
+
+import (
+	"errors"
+	"net"
+)
+
+//Anonymize struct that hold mask for V4 and V6.
+type Anonymize struct {
```
Comment (fho) [inline anonymizer.go]:
```suggestion
//IPAnonymizer anonymizes IP-Addresses by zeroing the Host part
type Anonymize struct {
```

File: anonymizer.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
+package goanonymizer
+
+import (
+	"errors"
+	"net"
+)
+
```
Comment (fho) [inline anonymizer.go]:
```suggestion

const DefaultIPv4Mask = "255.255.255.0"
const DefaultIPv6Mask = ""ffff:ffff:ffff:ffff:0000:0000:0000:0000""
```

File: anonymizer.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+	ipV6 := net.ParseIP(v6Mask)
+	if ipV6 == nil {
+		ipV6 = net.ParseIP("ffff:ffff:ffff:ffff:0000:0000:0000:0000")
+	}
+
+	return &Anonymize{v4Mask: net.IPMask(ipV4.To4()), v6Mask: net.IPMask(ipV6)}
+}
+
+//AnonymizeIp gets IP and transform it based on v4Mask and v6Mask.
+func (a Anonymize) AnonymizeIp(ipStr string) (string, error) {
```
Comment (fho) [inline anonymizer.go]:
```suggestion
// Anonymize anonymizes an IP-Address by setting the host-part to 0.
func (a *IPAnonymizer) Anonymize(ip string) (string, error) {
```

File: anonymizer.go (hunk 5 above)
Comment (fho) [inline anonymizer.go]:
```suggestion
	}
	
```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
+# GO IP anonymizer
+
+This library is built to give users more privacy by masking their IP address.
```
Comment (fho) [inline README.md]: Please also describe in the README how ip addresses are anonymized

File: README.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	newIP6, err := goanonymizer.NewAnonymize("","").AnonymizeIp(ip6)
+	if err != nil{
+		log.Println(err)
+	}
+
+	log.Printf("%s is transformed to %s", ip4, newIP4) //192.168.1.12 is transformed to 192.168.1.0
+	log.Printf("%s is transformed to %s", ip6, newIP6) //bbd1:e95a:adbb:b29a:e38b:577f:6f9a:1fa7 is transformed to bbd1:e95a:adbb:b29a::
+
+}
+
```
Comment (fho) [inline README.md]:
```suggestion
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,36 @@
+# GO IP anonymizer
```
Comment (fho) [inline README.md]:
```suggestion
# Go IP Anonymizer
```

File: README.md [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+This library is built to give users more privacy by masking their IP address.
+
+IPv4 and IPv6 addresses are supported. 
+
+Default subnet for IPv4 is `/24` and for IPv6 is `/64`. Those can be changed when library is initialized.
+
+##Example
```
Comment (fho) [inline README.md]:
```suggestion
## Example
```

File: anonymizer.go (hunk 6 above)
Comment (fho) [reply anonymizer.go]: To instantiate a default anonymizer, too empty strings needs to be passed to `New`. That is not soo elegant.

To make it easier to instantiate an IPAnonymized with the defaults, we could either:
- add a new constructor with no arguments
- define a package var like  `var DefIPAnonymizer = New(DefaultIPv4Mask, DefaultIPv6Mask)`

**Review by fho — COMMENTED** [review]
Please add:
- a go.mod file
- a LICENSE file

File: anonymizer.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+	ip := net.ParseIP(ipStr)
+	if ip == nil {
+		return "", errors.New("IP is not valid")
+	}
+	if ipv4 := ip.To4(); ipv4 != nil {
```
Comment (FranciscoKurpiel) [inline anonymizer.go]: I think a type switch reads a lot better:
```go
switch v := ip.(type) {
    case ipv4:
        return ...
    case ipv6:
        return ...
}
```

File: anonymizer.go (hunk 9 above)
Comment (FranciscoKurpiel) [inline anonymizer.go]: We are forcing the consumers of the library to use strings. I think we should split this function in two:
```go
func (a *IPAnonymizer) Anonymize(ip net.IP) (net.IP, error) {
    ...
}

func (a *IPAnonymizer) AnonymizeString(ipString string) (string, error) {
    // call a.Anonimize
}
```

File: anonymizer.go (hunk 6 above)
Comment (FranciscoKurpiel) [reply anonymizer.go]: @fho, what about the "options" approach?
```go
defAn := goanonimizer.NewAnonimizer()
noAn := goanonimizer.NewAnonimizer(
    goanonimizer.WithIPV4Mask(9), // mask out the last 9 bits of IPv4 addresses
)
```

Comment (fho) [reply anonymizer.go]: I think it's a lot of overhead here. 
I don't see that we will have more options in the future.
Passing both parameters as args is very simple.

Comment (fho) [reply anonymizer.go]: btw currently v4Mask and v6Mask seem to only support to be a full network mask, only passing the number of bits that describe the network should not work.
We could change it to support both notations

File: anonymizer.go (hunk 14 above)
Comment (fho) [reply anonymizer.go]: does not work, it's the same type :-|

File: anonymizer.go [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+	ipV6 := net.ParseIP(v6Mask)
+	if ipV6 == nil {
+		ipV6 = net.ParseIP("ffff:ffff:ffff:ffff:0000:0000:0000:0000")
+	}
+
+	return &Anonymize{v4Mask: net.IPMask(ipV4.To4()), v6Mask: net.IPMask(ipV6)}
+}
+
+//AnonymizeIp gets IP and transform it based on v4Mask and v6Mask.
+func (a Anonymize) AnonymizeIp(ipStr string) (string, error) {
```
Comment (fho) [inline anonymizer.go]: Using the same method for IPv4 and IPv6 addresses does not work as expected.
If an Ipv6 address is in the IPv4-mapped IPv6 address range, `To4()` will convert the Ipv6 address to an IPv4 address.

It also does not work to first call `To16()` and only `To4()` if To16 returns nil,
`To16()` converts IPv4 addresses to their IPv6 representation.

I'll add a method per IP version instead

**Comment by fho** [comment]
@FranciscoKurpiel @nocive I refactored the code according to our review and renamed the repository to `go-ip-anonymizer`.

Please review again

100% test coverage, btw! :-)
```
[fho@ltop goanonymizer]$ go test -cover  ./...
ok  	github.com/simplesurance/goanonymizer/ipanonymizer	0.002s	coverage: 100.0% of statements
```

File: ipanonymizer/anonymizer.go [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+package ipanonymizer
+
+import (
+	"errors"
+	"net"
+)
+
+var (
+	defaultIPv4Mask = net.IPv4Mask(255, 255, 255, 0)                                                     // /24
+	defaultIPv6Mask = net.IPMask{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0, 0, 0, 0, 0, 0, 0, 0} // /64
```
Comment (FranciscoKurpiel) [inline ipanonymizer/anonymizer.go:11]:
```suggestion
	defaultIPv4Mask = net.CIDRMask(24, 32)
	defaultIPv6Mask = net.CIDRMask(64, 128)
```

**Review by FranciscoKurpiel — APPROVED** [review]
Great implementation! Just a small optional comment.

File: README.md [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+```go
+package main
+
+import (
+    "fmt"
+    "net"
+    "os"
+
+    "github.com/simplesurance/goanonymizer/go-ip-anonymizer"
```
Comment (nocive) [inline README.md]: shouldn't it be?

```suggestion
    "github.com/simplesurance/go-ip-anonymizer/go-ip-anonymizer"
```

**Review by nocive — APPROVED** [review]

File: ipanonymizer/anonymizer.go (hunk 16 above)
Comment (fho) [reply ipanonymizer/anonymizer.go:11]: CIDRMask is easier to comprehend (if you already know what the args mean).

The current way is the same then how package vars are initialized in the `net` package and is more efficient.
These are also private variables and a user of the package never has to read them.
I keep the current variant. :-)

File: README.md (hunk 17 above)
Comment (fho) [reply README.md]: Yes, thanks

**Comment by fho** [comment]
branch was merged manually to master

**Pull request closed without merging**